src/core/cells.cpp

```diff
@@ -338,35 +332,52 @@ ParticleList sort_and_fold_parts(const CellStructure &cs, CellPList cells) {
 
   return displaced_parts;
 }
+} // namespace
 
 void cells_resort_particles(int global_flag) {
-
   invalidate_ghosts();
 
   clear_particle_node();
   n_verlet_updates++;
 
-  ParticleList displaced_parts =
-      sort_and_fold_parts(cell_structure, cell_structure.local_cells());
+  static std::vector<Cell *> modified_cells;
+  modified_cells.clear();
+
+  ParticleList displaced_parts = sort_and_fold_parts(
+      cell_structure, cell_structure.local_cells(), modified_cells);
+
+  for (auto const &p : displaced_parts) {
+    cell_structure.update_particle_index(p.identity(), nullptr);
+  }
 
   switch (cell_structure.type) {
   case CELL_STRUCTURE_NSQUARE:
-    nsq_exchange_particles(global_flag, &displaced_parts);
+    nsq_exchange_particles(global_flag, &displaced_parts, modified_cells);
     break;
   case CELL_STRUCTURE_DOMDEC:
-    dd_exchange_and_sort_particles(global_flag, &displaced_parts, node_grid);
+    dd_exchange_and_sort_particles(global_flag, &displaced_parts, node_grid,
+                                   modified_cells);
     break;
   }
 
+  boost::sort(modified_cells);
+  for (auto cell : modified_cells | boost::adaptors::uniqued) {
```

Comment:
Bash syntax in C++ code. What a time to be alive :smile:

Reply:
With C++20 all code will look like this :-)

Reply:
I looked up the Boost docs on Ranges and found how to pipe data to grep-expressions and sed-expressions, Bash style:
```c++
#include <boost/range/adaptor/transformed.hpp>
#include <boost/range/adaptor/filtered.hpp>
#include <boost/range/algorithm/copy.hpp>
#include <iterator>
#include <iostream>
#include <vector>
#include <string>
#include <regex>

struct grep_expression {
  bool operator()(std::string const &s) const {
      return std::regex_match(s, std::regex("[A-Z]{2}.*"));
  }
};

struct sed_expression {
  std::string operator()(std::string const &s) const {
    return std::regex_replace(s, std::regex("ESPResSo"), "CAPPucciNo");
  }
};

int main() {
    using namespace boost::adaptors;
    std::vector<std::string> input = {"ESPResSo", "Python", "LB"};
    boost::copy(
        input | filtered(grep_expression()) | transformed(sed_expression()),
        std::ostream_iterator<std::string>(std::cout, "\n"));
    return 0;
}
```